docs/tutorials/gallery_adding_components.py

```diff
@@ -0,0 +1,125 @@
+"""
+# Adding Components to Ragna
+
+Ragna has builtin support for several [assistants][ragna.assistants] and 
+[source storages][ragna.source_storages], but there may be cases where you 
+want to use one that is not currently supported.
+
+This tutorial walks you through the basics of adding components
+that are not currently officially supported. 
+"""
+
+# %%
+# ## Adding an LLM Assistant
+
+from typing import Iterator
+
+from ragna.core import Assistant, Source
+
+
+class TutorialAssistant(Assistant):
+    def answer(self, prompt: str, sources: list[Source]) -> Iterator[str]:
+        yield (
+            f"This is a default answer. There were {len(sources)} sources."
+            ""
+            f"The prompt was"
+            f"{prompt}"
+        )
+
+
+# %%
+# The main thing to do is to implement the [`answer()`][ragna.core.Assistant.answer] abstract method.
+# The [`answer()`][ragna.core.Assistant.answer] method is where you put the logic to access your LLM.
+# This could call an API directly, call other member functions of your assistant that call an API,
+# or call a local LLM. Ragna is designed to give you that flexibility.
+
+# %%
+# Your [`answer()`][ragna.core.Assistant.answer] method should take a prompt in the form of a
+# string, and a list of [`Source`][ragna.core.Source]s, in addition to whatever other arguments
+# necessary for your particular assistant. The return type is an [`Iterator`](https://docs.python.org/3/library/stdtypes.html#typeiter) of strings.
+
+# %%
+# !!! note
+#     Ragna also supports streaming responses from the assistant. See the
+#     [example how to use streaming responses](../../generated/examples/gallery_streaming.md)
+#     for more information.
+
+# %%
+# ## Adding a Source Storage
+
+from ragna.core import Document, Source, SourceStorage
+
+
+class TutorialSourceStorage(SourceStorage):
+    def __init__(self):
+        # import database api here
+
+        # set up database
+        self._storage: dict[int, list[Source]] = {}
+
+    def store(self, documents: list[Document], chat_id: int) -> None:
```

Comment:
There are three things here:

1. The `chat_id` is not an `int`, but rather a `uuid.UUID`.
2. So far we have not explained what `chat_id` is and where it comes from. It is not part of the standard API:

    https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L102-L103

    https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L111-L112

    This deserves a full section below. Meaning at this point, we need to tell the reader "we need a way to identify the chat that we are storing documents for. We can use the unique `chat_id` for this. To get access to it, just add it as parameter to method signature. We'll explain how this works in section X".
3. It is convention to separate the non-standard parameters to be keyword-only.
    ```suggestion
        def store(self, documents: list[Document], *, chat_id: int) -> None:
    ```

Reply:
Re 2.: Here is how it works:

When adding new parameters to the signature of the protocol methods

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L99-L100

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L210-L213

we automatically pick up on this and add the parameters to the model of the component

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L73-L76

When creating a new `Chat`, we check if the additional parameters satisfy the additional parameters of the selected components:

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_rag.py#L251

You don't need to understand this in detail. The mental model that you can use is that the extra `params` that you can pass to the `Chat`

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_rag.py#L77-L84

are automatically checked for completeness and distributed to the components at runtime. Meaning, every parameter you define without a default, must be passed when creating the chat. If you add a default value, this can be overridden by passing a value to the chat, but you don't have to.

So why didn't we need to pass the `chat_id` although it didn't have a default? `chat_id` is a special parameter alongside `chat_name` and `user`:

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_rag.py#L103-L108

They are automatically set by Ragna and can always be added to a signature without the need for the user to supply them. Same as the other parameters they can be overridden by passing other values when creating the chat. This is what we do for the REST API:

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/deploy/_api/core.py#L204-L207